### DISCUSSION TITLE
Layers

### DISCUSSION THREAD

**Discussion by joshka** [cbeck88/conf-rs#1] [General]
Hey, just wanted to drop a comment about the motivation doc. That write up was really excellent! I'm convinced this is the right approach for config for many things. I'm writing a new web app that's intended to be installed by end users (e.g. delivered as a dockerized app and configured). I'd like this to be easily able to be configured with command line (`docker run -- ...`), environment vars, and plain old config files. It seems like this library comes close to solving that that problem, but omits loading configs from files. This seems like something desirable to add in this sort of situation, and I wonder if you could expand on your thoughts a bit there.

I do note that you do have a mechanism for loading from files, but the issue is that this doesn't seem to easily integrate with the required values in a way that would explain where the values are set etc.

I've used figment + clap a few times and run into a various sets of problems similar and different than yours - mostly in how exactly to express optional values in one layer, but having the overall value being required somewhere in the layer of values. Configuring `clap` to handle this basically entails making a parallel clap options struct where all the parameters are optional. This is kinda messy and doesn't scale nicely.

One thing that I did experiment with recently was having 3 layers in figment - clap, toml config file, and defaults. In the web app I was able to present a page that showed these layers of values as keys and values. This approach seems useful in the case of writing a web app that is a product (as opposed to where the webapp is the development activity of an organization). The page becomes a sort of answer to two questions simultaneously "how is this configured", "why is this configured with that value?". Does this seem like something that could be handled by conf (with changes obv.)?

More generally, if you think about how config systems are documented, it's fairly common to have a doc that lists out all the config properties, their command line aliases, their identifiers, their defaults and to provide information on setting them. I see this crate could be a really good way to generate that doc in a way that would always be in sync with the allowed values.

TL;DR: My ideal config system is very close to what you're doing with conf, but with
- load from cli args
- load from env vars
- load from file based on cli arg (e.g `--config`)
- load from file based on default location
- load from default file
- load from struct defaults (MyConfig::default()

I like everything about how conf works except for those missing middle pieces.

**Comment by cbeck88** [comment]
Thanks for your thoughts and feedback!

I think I had a vague idea along the following lines:

* There is a function `Conf::try_parse_from`: https://github.com/cbeck88/conf-rs/blob/2fb6bfe011b5c3a8471c6c0b3c59505deed2da82/src/traits.rs#L50
* Create an alternate version of this where, a key-value pair can also be passed with "provenance". TBD what the API of that is but it should basically indicate e.g. "this value came from this config file", "this value came from env", etc., and carry enough info to be rendered usefully in help / error messages.
* We're already kind of tracking that in conf using the `ValueSource` enum, so likely we'd create another value in that enum where "user-specified" provenance info can be stored and then used in error messages.
* Then, "layers" can sort of get out-sourced to a preprocessing step that happens before `try_parse_from`.

In my mind, a preprocessing step is attractive in part because, if the user does want to merge a bunch of maps from many sources together, they can make it work exactly how they want, and this concern can be mostly separated from the code that already exists in conf. This is somewhat more sophisticated than outsourcing this concern to a `dotenv` crate or similar, and means that you don't lose the provenance of the values.

One issue is that, if this is strictly preprocessing, then the exact thing you are asking for is made more difficult:

> load from cli args
load from env vars
load from file based on cli arg (e.g --config)
load from file based on default location

If we don't know all of the input values until after we've seen the `--config` option in the args, then we can't call `try_parse_from` until after we've parsed some args, so we'd likely have to parse the args twice. Really you'd have to just find the `--config` flag and its argument first. It's not impossible but it's more complicated for sure. On the other hand, if it's like, "load from file based on CONFIG_FILE path", then you can just check std::env for that first, then check for the file existing and load any values from it, and forward to `try_parse_from`.

---

On `clap` itself there is an old ticket about this, where he's suggesting some kind of plugin system (dependency injection) rather than thinking of this as a preprocessing step: https://github.com/clap-rs/clap/issues/4793

I am not sure if that kind of approach would actually let you do the thing you are suggesting more simply. If all the plugin system does is give you a place to stick a lambda function, you can already kind of do that with `value_parser`. If the idea is that it should give you access to the intermediate results of parsing (so, look at the `--config` parameter while parsing other parameters), I'm not sure how it would work because you might not have read that parameter yet.

---

I think if I had to do exactly what you are saying with `--config`, what occurs to me is like

1. Make a helper struct that parses that value:

```
struct ValueSourceConfig {
   #[arg(long)]
   pub config: Option<PathBuf>
}
```

2. In `fn main()`, do `ValueSourceConfig::parse()` first, then look for a file.
3. Merge that with values in `std::env::vars()`, and pass to `MyConfig::parse_from`.


In step 2, you might have a problem where `ValueSourceConfig::parse` chokes on unrecognized flags. In `clap` there is a setting [`ignore_errors`](https://docs.rs/clap/latest/clap/struct.Command.html#method.ignore_errors) that I think will usually allow it to succeed anyways. For `conf` I don't think I'm exposing something like that, we could think about how it might work.

Another problem is, either library would likely catch the `--help` option and bail, but that's not what you want here. I think they both have disable_help options though.

To avoid weird interactions, what I would probably do is, if you really don't think `ValueSourceConfig` will get more complicated, just do that parser manually by iterating over `std::env::args` or similar. Or you could try to use one of the really minimal arg parse libraries, like pico-args or something (see also: https://github.com/rosetta-rs/argparse-rosetta-rs), just to get that part.

If it were my project, I would be asking myself why I have to support `--config` to specify the config file path, and can't just ask the users to set `CONFIG` env.

Let me know what you think about all that, thanks again!

**Reply by joshka** [reply]
Thanks for the great reply. Lots to process and follow up reading there. I may not reply for a bit as the various bits sink in.
In particular seeing that clap does have effectively what I want in a system on their radar is good to see. https://github.com/clap-rs/clap/discussions/3476#discussioncomment-2185709 seems pretty close to what I'm looking for.

The `--config` is mostly a convenience thing. `docker run foo/bar --config baz.toml` is a bit easier and often more discoverable than `docker run foo/bar -e CONFIG_PATH=baz.toml`. In particular this can participate in shell completion. Ack on this being a difficult thing to get right with respect to parsing order though.

**Reply by cbeck88** [reply]
I haven't actually *used* figment, but as I look at the docs, I wonder if I should try to work an example where you use `figment` to load env and load config files, then "extract" that into a `HashMap<String, String>` to just get all the merged env values, and then use that with `Conf::try_parse_from`.

That approach might be able to side-step a bunch of the issues with `serde::Deserialize` because if I understand right, deserializing from the figment to a HashMap basically won't fail.

**Reply by cbeck88** [reply]
I guess this isn't quite right, `figment` has a richer data model than that: https://docs.rs/figment/latest/figment/value/enum.Value.html

---

Maybe the right level of abstraction is something like, `Conf` gets an extension to `try_parse_from` that takes an arbitrary [`serde::Deserializer`](https://docs.rs/serde/latest/serde/de/trait.Deserializer.html), in addition to args and env, and try to use that thing as a fallback (but before the default value), when walking the struct and filling in the struct values. I need to ponder that some more and figure out if that makes sense. I think maybe if the `Conf` macro is in control, and not the `serde::Deserialize` trait, then it can manage all the error handling nicely, but still get the values properly out of the `Deserializer`.

**Reply by cbeck88** [reply]
I have been working through the details of this, and I think it’s at least plausible that it can work.

However, there are some things that I’m not sure about, and it would help if I can understand how your config files look and why. For example when you are using figment.

 1. When you have a large config structure composed of smaller config structures, do you want all that to be reflected clearly in your config files? Or do you use serde flatten to flatten it all down so that your config file is just key value pairs? It seems like, you might see some benefit in having some of the structure reflected in the file for organizational purposes, but in other cases, if you turn a field into a struct but want to avoid a breaking change, then you need to be able to make the serde flatten happen.
2. It’s a lot simpler if `conf` only supports the latter and not the former. However, then it’s not clear to me why you need the feature at all, you could probably do what I was suggesting and merge config file values with env yourself and then pass the merged result to `Conf::try_parse_from`. If we added a way to note provenance then I don’t see any major issues. Or, your conf file could just be a .env file. But, if that would meet your needs you probably wouldn’t have posted this discussion.
3. So, suppose we added a feature where you can parse from args, env, and “structured data” (from a config file, or possibly merged from several files) represented as a serde Deserializer. And the priority is args, then env, then config file, then any defaults. What serde Deserialize features are important here? Flatten? Maybe prefix? Rename individual fields?

**Reply by cbeck88** [reply]
I guess I'm not sure if my idea will actually work, what I was thinking was, where we currently have `Conf::from_conf_context`:

https://github.com/cbeck88/conf-rs/blob/2fb6bfe011b5c3a8471c6c0b3c59505deed2da82/src/traits.rs#L87

we could instead (or maybe in addition) do an implementation of `serde::DeserializeSeed`: https://docs.rs/serde/latest/serde/de/trait.DeserializeSeed.html

So it might look like

```
impl<'de> DeserializeSeed<'de> for ConfContext {
  type = YourStruct;
  fn deserialize(self, deserializer: ...) -> Result<YourStruct, De::Error> {
    ...
  }
}
```

And this would attempt to create the structure with config file data being walked using the serde framework, but with all the data from the `ConfContext` as well to fill in holes and to shadow the config file as necessary.

The problem with this is, the error handling is going to get messed up, because we can only return a `de::Error` coming from the deserializer library, and not the `conf::Error` type.

The reason I thought it might work is that, there is another difficult issue one confronts in the `MapAccess` trait:

https://docs.rs/serde/latest/serde/de/trait.MapAccess.html

If one is doing a custom implementation of `Deserialize`, eventually your visitor gets passed a `MapAccess`. What would be convenient is if I can recurse by getting a `Deserializer` instance for each child of the map. But the way the API works, it wants to actually put the whole `Deserialize` operation in this call to `next_value` that obtains the child value.

I thought about creating a helper struct that implements `Deserialize` by simply taking the passed `Deserializer` and storing it for later use. Then we can slow the process down and recurse the way I wanted. But because the `Deserializer` is not object-safe, it's not clear how you can actually implement that. You would need to be able to create a suitably type-erased `Deserializer` somehow but I can't see that it's possible.

Implementing `DeserializeSeed` on `ConfContext` allows us to avoid that issue, because then we can use `MapAccess` as intended. The problem then is that, we have to return a deserializer error rather than our own error type, and we can't really collect multiple errors in a good way, because the serde error trait doesn't support that.

---

So I'm thinking, taking a `serde::Deserializer` may be a dead end.

---

I'm thinking the most practical thing is probably:

* Only allow one, fixed, config document type, like `serde_json::Value`, or perhaps `figment::Value`, and implement something to allow using that as one of the value sources for the `Conf` structure.
* Do an optional extension that integrates with `figment`, so that if you want layering like this, you just pass in the `&Figment`, and then under the hood we extract it to a `figment::Value` and then walk that structure as needed to marshall it onto the `Conf` structure, at higher priority than the defaults, but lower priority than args and env.

**Reply by cbeck88** [reply]
Maybe it is possible to type-erase a `Deserializer` adequately. The trait has generic functions, but they are only generic over the Visitor type, and I always know what visitor types I'm going to use. I'm not sure that I ever actually have to pass the type-erased deserializer back to the `serde::Deserialize` trait or related -- if I don't then that idea might work out.

**Reply by cbeck88** [reply]
If I have some time later, I'm going to try to do a POC, we'll see if my mental model here is accurate, but I think it will work

**Reply by cbeck88** [reply]
I did a draft PR here: https://github.com/cbeck88/conf-rs/pull/7

It needs more tests and documentation, and there are more features that it should have to be configurable and feel cohesive.
I had to abandon my original idea, but I got another approach to work

**Reply by cbeck88** [reply]
This should now be considered done in version 0.2, we now have `serde` support with most of the serde attributes that are relevant.

**Comment by cbeck88** [comment]
> One thing that I did experiment with recently was having 3 layers in figment - clap, toml config file, and defaults. In the web app I was able to present a page that showed these layers of values as keys and values. This approach seems useful in the case of writing a web app that is a product (as opposed to where the webapp is the development activity of an organization). The page becomes a sort of answer to two questions simultaneously "how is this configured", "why is this configured with that value?". Does this seem like something that could be handled by conf (with changes obv.)?

I definitely like this idea -- I think it should be in scope to either always generate some kind of report like this, or provide some kind of hook or customization point that can be used to do it. I'm going to make an issue to follow up on this specifically
